src/Nest/XPack/MachineLearning/Job/Detectors/CountDetector.cs

```diff
@@ -0,0 +1,185 @@
+using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using Newtonsoft.Json;
+
+namespace Nest
+{
+	public class CountFunctions
```

Comment:
This is an interesting area (the next few comments are the same).

@russcam was thinking of maybe limiting this to an enum. We could also make it private. The value only really is used for deserialization.

Thoughts?

Reply:
They could be made `private`, but the implementation would need to change to accommodate because they are passed into the constructor of the abstract base class in each case e.g.

```c#
public abstract class CountDetectorBase : DetectorBase, ICountDetector
{
	public Field ByFieldName { get; set; }
	public Field OverFieldName { get; set; }
	public Field PartitionFieldName { get; set; }

	protected CountDetectorBase(string function) : base(function)
	{
	}
}

public class CountDetector : CountDetectorBase
{
	public CountDetector() : base(CountFunctions.Count) {}
}
```

Been discussing making these `enum`s so that someone creating a new instance of a descriptor cannot violate the API e.g.

```c#
public class CountDetectorDescriptor<T> : DetectorDescriptorBase<CountDetectorDescriptor<T>, ICountDetector>, ICountDetector where T : class
{
//...
	public CountDetectorDescriptor(string function) : base(function) {}
//...
}
```
This should be restricted only to accepting `CountFunctions.Count`, `CountFunctions.HighCount` or `CountFunctions.LowCount`. I think the base descriptor and OIS types should accept function as a `string` though, so proposal would look like

```c#
public class CountDetectorDescriptor<T> : DetectorDescriptorBase<CountDetectorDescriptor<T>, ICountDetector>, ICountDetector where T : class
{
//...
	public CountDetectorDescriptor(CountFunction function) : base(function.GetStringValue()) {}
//...
}
```